### ISSUE TITLE
VM should ensure that the last operation is a `HALT`

### ISSUE THREAD

**Issue by plafer** [0xMiden/miden-vm#1383]
The processor should ensure to always insert a `HALT` (even when no padding is needed). We should also have a boundary constraint that ensures that the last statement is a `HALT`.

This is because our multiset checks (buses and virtual tables) assume that no multiset element can be generated/removed from the last trace row (and hence we need to ensure that the last operation in the trace is an `END`).

**Comment by bobbinth** [comment]
Good point! I thought we already had this - but I guess since decoder constraints are not yet implemented in Rust, we didn't add any assertions against the decoder part of the trace.

**Comment by plafer** [comment]
This is fixed, and a duplicate of #2251
